numpy/core/src/multiarray/item_selection.c

```diff
@@ -1749,6 +1750,324 @@ PyArray_LexSort(PyObject *sort_keys, int axis)
+    aelsize = (npy_intp)PyArray_ITEMSIZE(self);
+    maxelsize = PyArray_DESCR(mps[0])->elsize;
+    needcopy = (rstride != sizeof(npy_intp));
+    for (j = 0; j < n; j++) {
+        needcopy = needcopy
+            || PyArray_ISBYTESWAPPED(mps[j])
+            || !(PyArray_FLAGS(mps[j]) & NPY_ARRAY_ALIGNED)
+            || (PyArray_STRIDES(mps[j])[axis] != (npy_intp)PyArray_DESCR(mps[j])->elsize)
+            || (selstride != sizeof(npy_intp)) ;
+        if (PyArray_DESCR(mps[j])->elsize > maxelsize) {
+            maxelsize = PyArray_DESCR(mps[j])->elsize;
+        }
+    }
+
+    if (needcopy) {
+        char *valbuffer, *indbuffer;
+        int *swaps;
+        assert(N > 0);  /* Guaranteed and assumed by indbuffer */
+        npy_intp valbufsize = N * maxelsize;
+        if (NPY_UNLIKELY(valbufsize) == 0) {
+            valbufsize = 1;  /* Ensure allocation is not empty */
+        }
+
+        valbuffer = PyDataMem_NEW(valbufsize);
+        if (valbuffer == NULL) {
+            goto fail;
+        }
+        indbuffer = PyDataMem_NEW(N * sizeof(npy_intp));
+        if (indbuffer == NULL) {
+            PyDataMem_FREE(valbuffer);
+            goto fail;
+        }
+        swaps = malloc(NPY_LIKELY(n > 0) ? n * sizeof(int) : 1);
+        if (swaps == NULL) {
+            PyDataMem_FREE(valbuffer);
+            PyDataMem_FREE(indbuffer);
+            goto fail;
+        }
+
+        for (j = 0; j < n; j++) {
+            swaps[j] = PyArray_ISBYTESWAPPED(mps[j]);
+        }
+        while (size--) {
+            iptr = (npy_intp *)indbuffer;
+            for (i = 0; i < N; i++) {
+                *iptr++ = i;
+            }
+            for (j = 0; j < n; j++) {
+                int rcode;
+                elsize = PyArray_DESCR(mps[j])->elsize;
+                astride = PyArray_STRIDES(mps[j])[axis];
+                argsort = PyArray_DESCR(mps[j])->f->argsort[NPY_STABLESORT];
+                if(argsort == NULL) {
+                    argsort = npy_atimsort;
+                }
+                _unaligned_strided_byte_copy(valbuffer, (npy_intp) elsize,
+                                             its[j]->dataptr, astride, N, elsize);
+                if (swaps[j]) {
+                    _strided_byte_swap(valbuffer, (npy_intp) elsize, N, elsize);
+                }
+                rcode = argsort(valbuffer, (npy_intp *)indbuffer, N, mps[j]);
+                if (rcode < 0 || (PyDataType_REFCHK(PyArray_DESCR(mps[j]))
+                            && PyErr_Occurred())) {
+                    PyDataMem_FREE(valbuffer);
+                    PyDataMem_FREE(indbuffer);
+                    free(swaps);
+                    goto fail;
+                }
+                PyArray_ITER_NEXT(its[j]);
+            }
+            char *slicebuffer = PyDataMem_NEW(N*aelsize);
+            _unaligned_strided_byte_copy(slicebuffer, aelsize, sit->dataptr,
+                                         selstride, N, aelsize);
+            npy_intp dims[1] = {N};
```

Comment:
I have to admit, I am curious if we cannot just replace the below code with a manual `memmove`, you have to move elements (`elsize`), we know that every item ends up in exactly one other place, and we do not have to check the indices for out-of-bound values...

so a simple for loop with a `memmove` inside could be just as well.  If it turns out a bit slow for some (smaller) dtypes, we could take the approach we have in `npy_fastputmask` to have the compiler generate different versions for the most common dtype sizes.

After doing that, we could experiment with an in-place algorithm, which seems like it should be possible as long as we also modify `indbuffer` at the same time.

In any case, right now I think a simple for-loop with a `memmove` inside is probably just as fast (no overhead and no index checking) or aster. And probably even slightly easier to read.

Reply:
Ah will, look into this.